text/0753-logging-and-debugging-context-api.md

```diff
@@ -0,0 +1,189 @@
+  such parents, then it is an array of objects. Otherwise, it is `null`.
+
+Debugging context parents are the same parents that are associated with the
+object via `associateDestroyableChild`. If a particular parent exists, but does
+not have a debugging context generator, then it is skipped and its parent(s) are
+used instead.
+
+Debugging context generators for a given object are set via
+`setDebuggingContextGenerator`. When you call `getDebuggingContext` on a
+particular object, it first checks to see if a generator was associated directly
+with that object. If one wasn't, it then checks the prototypes of that object.
+This way, debugging context generators can be associated with classes instead of
+individual objects. When one is found, it is called and passed the object that
+`getDebuggingContext` was called on.
+
+### setDebuggingContextGenerator
+
+`setDebuggingContextGenerator` associates a generator function with a given
+object. This generator function is called when `getDebuggingContext` is called
+on that object, or if the object is a class, then when it is called on an
+instance of the class. The generator is inherited, so subclasses will get the
+same generator unless it is overridden.
+
+The generator function receives the object itself as its only parameter. It
+should return an object with the following properties:
+
+- `description`: A helpful string describing the object for debugging purposes.
+- `extra`: This is an optional property which can contain extra debugging
+  information. It can be any type of value.
+
+## How we teach this
+
+This API is meant to be used by library authors and framework authors, so it
+does not need a dedicated section in the guides yet. In time, a section could be
+added in the In-Depth Topics section once best practices are established.
+
+In the meantime, the descriptions of the new functions from the detailed design
+section above can be used as the basis for API documentation.
+
+## Drawbacks
+
+The main drawback is that this API cements the need for a child-to-parent link
+in destroyables, as it is necessary to be able to find the full debugging
+context at any time for a given object. This was not specified directly by the
+previous RFC, but it is necessary to prevent memory leaks, since children can be
+destroyed at any time, and if they are not removed from their parents then they
+will leak.
+
+## Alternatives
+
+- We could use a different import path than `@glimmer/debug`. This could be
+  useful, as we may want to reserve that namespace for APIs like `assert` and
+  `deprecate` which _will_ be stripped from production, and co-mingling stripped
+  and non-stripped APIs could be confusing.
+
+- We could use a separate tree specifically for debugging context/information.
+  This would likely be very costly from a performance perspective, and would
+  likely be prohibitive.
+
+- We could only ship this API in development and not production, relaxing the
```

Comment:
could there be an option to opt in to stripping this or other apis(?) out only in production?

Reply:
I think the cost here would be minimal enough that it won't really matter, and I think there could be broader repercussions to consider if we allowed that. For instance, some apps/addons may assume it exists, and stop working if it does not. I also worry about increasing implementation cost, and increased cost for Embroider to add another config option. That said, I'm open to the idea, would definitely like to see what others think about it.